### ISSUE TITLE
Lossy webp support

### ISSUE THREAD

**Issue by DVLP** [matanui159/jebp#1]
Great idea with one-file stb approach for webp! Obviously adding lossy would make this so much better. This library will be a very useful tool for people who want to make lean WASM builds.

**Comment by matanui159** [comment]
Yeah, been bike shedding a little bit but plan to start working on lossy support soon:tm:

As for wasm, I did consider adding SIMD-detection for it and adding it to the tested architectures but since Safari now supports WebP the need for a small wasm webp decoder is probably going to be dying soon

**Comment by DVLP** [comment]
This may be true for browsers but not for node.js. Also good to have a way to read webp in a worker without accessing canvas

**Comment by DVLP** [comment]
EDIT: Never mind these were just wrong array sizes - uint8_t instead of uint32_t for targets

I have a version of libwebp with most of the encoding stuff removed so it takes half the space but still I think your project will be much smaller no matter how many files I remove. I thought making a smaller version will solve my problem with libwebp working strangely. i.e. it gives wrong dimensions for images with WebPGetInfo. Is libwebp also sometimes not starting for you or not working after a page refresh?

Here's the project in Emscripten, feel free to take any code. Just extracted it from my large project to create an environment to debug the problem in isolation.
https://github.com/DVLP/webp-parser

To see how it reports dimensions incorrectly:

1. git clone https://github.com/DVLP/webp-parser.git
2. npm i
3. npm start
4. Use file selection input to pick up any WEBP file and click on Parse button. In the console it will return an array with 3 values.

**Comment by matanui159** [comment]
I have opened a draft PR to track progress: #2

**Comment by DVLP** [comment]
Great, I'm looking forward to use it one day.

**Comment by DVLP** [comment]
What's the performance you've been getting so far compared to libwebp? I tried compiling the lossy branch in Emscripten but getting lots of errors(btw emcc uses llvm/clang). Mostly it's due to casting of types. I added necessary casts and it managed to compile, no luck with decompression yet. For now I can make a PR with these to make the code more compiler agnostic

**Comment by matanui159** [comment]
Currently unfinished. Have taken a bit of a break but probs gonna come back to take a break from other projects.
Does Emscripten/Clang have those warnings by default? I used to have warnings enabled for type conversions (`-Wconversion`) but found it too hard making the types consistent enough so I scrapped those changes.
I have looked into Emscripten support. If I did add it, it would probably be with a JavaScript wrapper and API, but currently not a lot of drive to look into it now that Safari supports WebP.

**Comment by matanui159** [comment]
The lossy branch should be in a mostly usable state at this point. Its still missing deblocking filters but if you're fine with the reduced quality by not having those, the branch has been tested and does seem to work on a few test images.

**Comment by DVLP** [comment]
Thanks for an update! I managed to decode one file, but most files I have are using VP8X (0x58385056 tag), is this where the alpha channel is? From what I see in libwebp VP8X can contain all kinds of data but after it there should be normal VP8(L)

Btw I needed to add some modifications to make it compile
This markup is not supported:
`jebp__huffman_group_t *groups = &(jebp__huffman_group_t){0};`

Another place that I needed to change is where "goto" are  i.e. `goto free_groups`. It needed all variable definitions like `jebp_int nb_read_groups = 0;` to be moved above goto to compile.

**Comment by matanui159** [comment]
Are you compiling with C99? Both of those should be standard C99 syntax 🤔 

Also VP8X is another thing seperate to VP8 support, so would be done after this is done and merged

**Comment by DVLP** [comment]
It's emscripten and default C++ with no extensions. Emscripten uses clang. What does this do?
`= &(jebp__huffman_group_t){0};`
In SO I can only find that it's an error https://stackoverflow.com/questions/63130256/simplest-fix-for-error-taking-the-address-of-a-temporary-object
Actually this and the goto's seem to be easy to change and with these changes the code becomes much more portable

**Comment by matanui159** [comment]
Well... it's standard _C_, ig not standard C++ :/ I've been coding against C99 but I probs should allow support for C++ as well

For reference though, they're called [compound literals](https://en.cppreference.com/w/c/language/compound_literal):
> The value category of a compound literal is _lvalue_ (its address can be taken).
> [...]
> in which case the object's _lifetime_ ends at the end of the enclosing block

**Comment by matanui159** [comment]
I might look into seeing how difficult it would be to switch to C89 and potentially start using stricter `-Wconversion`, although these would be done as their own work initially on the main branch and then extended to this branch (or after this branch is done).
